### PR TITLE
fix(interactions): improve fallbacks in resolved data

### LINKED ISSUE #644: Process crashing without any intents
### Summary

When using `Intents.none()`, occasionally the bot will crash with the following [error](https://paste.disnake.dev/?id=1658214731948219)

### Reproduction Steps

Turned on my bot with `intents=Intents.none()`, while the bot is running it then proceeds to crash occasionally with the above error. 

After some back n forward we believe this error is due to guilds the bot has joined during runtime, as it doesn't have any runtime information on the guild itself to use.

You can read the entire discussion [here](https://discord.com/channels/808030843078836254/942319505915412500/995709583173627934)

### Minimal Reproducible Code

_No response_

### Expected Results

My bot not to crash.

### Actual Results

My bot crashed.

### Intents

`Intents.none()`

### System Information

```markdown
Python3.10
Disnake master
```

### Checklist

- [X] I have searched the open issues for duplicates.
- [X] I have shown the entire traceback, if possible.
- [X] I have removed my token from display, if visible.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by shiftinv**
## Summary

Improves channel/guild fallbacks in resolved interaction data, also fixing typing issues.
Resolves #644.

## Checklist

- [x] If code changes were made, then they have been tested
    - [ ] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [x] This PR fixes an issue
- [ ] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Review by onerandomusername — APPROVED** [review]
This is mostly good, although there is still a minor somewhat-bug, but I think this out of scope of this pull request.

With a command defined as 
```py
async def noguild(self, inter: disnake.CommandInteraction, channel: disnake.abc.GuildChannel) -> None:
```
the results of the variables are as follows:
```py
>>> inter.channel.guild
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: 'PartialMessageable' object has no attribute 'guild'
>>> channel.guild
<Object id=755083284416954428>
>>> inter.guild
None
```

**Comment by shiftinv** [comment]
@onerandomusername Good point. Should keep that in mind, as far as I can tell that would potentially be fixed by #647

File: changelog/646.bugfix.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Improve channel/guild fallback in resolved interaction data.
```
Comment (onerandomusername) [inline changelog/646.bugfix.rst:1]: Could we explain how this was fixed and what changes this will bring to a user?

Comment (shiftinv) [reply changelog/646.bugfix.rst]: 2d755778e942cb465fcf9ba25a390448de9ed98c

**Pull request merged**